### PR TITLE
[Model][DiffusionGemma] Bound sampler buffer memory under serving pressure

### PR DESCRIPTION
**Pull request by masterFoad**
Draft PR for maintainer feedback.

## Summary

DiffusionGemma decode can create full-vocab sampler buffers sized by sampler rows times vocabulary after KV-cache sizing. At serving pressure, that temporary allocation can OOM even when model loading and KV allocation succeed.

This PR adds an opt-in `row_chunked` DiffusionGemma sampler path. In that path, `compute_logits()` returns hidden states and the sampler projects a bounded number of rows at a time. The dominant live full-vocab sampler term changes from `O(R * V)` to `O(K * V)`, where:

```text
R = num_decode_requests * canvas_length
V = vocab_size
K = row_chunk_size
```

Default behavior is unchanged. The current materialized sampler remains the default.

```bash
VLLM_DIFFUSION_GEMMA_SAMPLER_BACKEND=row_chunked
```

The branch also includes a default-no-op model hook for model-reported non-KV runtime memory reserve, applies that reserve during GPU KV-cache sizing, and falls back to the existing materialized sampler when logprobs are requested. I can split the reserve hook if maintainers prefer that as a separate change.

## Implementation notes

- Avoid pre-sampler full-vocab logits in the opt-in path: `compute_logits()` returns hidden states for `row_chunked` instead of materializing `[sampler_rows, vocab]` logits before sampling.
- Bound live sampler rows: the rowchunk sampler projects only `row_chunk_size` rows at a time, changing the dominant live full-vocab term from `O(R * V)` to `O(K * V)`.
- Preserve chunk-size invariance: non-greedy sampling uses stateless row/token-indexed randomness, so the sampled token for a row is independent of row chunk size.
- Bound random-sampling scratch: Gumbel generation is tiled over vocab inside each row chunk.
- Reduce per-chunk overhead: the helper reuses `log_softmax` normalization results and hoists row-local RNG setup outside the inner vocab-tile loop.
- Keep logprobs safe: requests that need logprobs use the existing materialized sampler path.

Only the dominant shape change is quantified: with `R=4096` and `K=64`, the live full-vocab row term is `R / K = 64x` smaller. The measured peak numbers below are for the combined rowchunk path; the smaller optimizations are not separately attributed.

## Related work

- #45672 is the closest overlap. It bounds DiffusionGemma sampler transient memory by tiling request groups inside the materialized sampler path.
- I measured the two shapes locally. Request tiling is effective for sampler transient when the request group is small. The architectural difference is that request tiling still starts from full materialized logits, while this branch avoids that full logits tensor in the opt-in path.
- I am happy to contribute the validation harness or rework this as a follow-up if maintainers prefer #45672 as the base direction.
- #45689 is relevant to the logprobs fallback path.
- #45719 is relevant DiffusionGemma multi-GPU context. TP/PP support is outside this PR.
- #37518 is related to sampler memory accounting and KV sizing.
- #45369 uses a similar full-vocab buffer reduction pattern in another sampler path.

## Test plan

Local checks:

```bash
ruff check <changed files>
.venv/bin/python -m py_compile <changed python files>
git diff --check
.venv/bin/python -m pytest tests/model_executor/test_diffusion_gemma_rowchunk_sampler.py -q
.venv/bin/python -m pytest tests/v1/worker/test_diffusion_sampler_memory_reserve.py -q
.venv/bin/python -m pytest tests/v1/worker/test_model_memory_reserve.py -q
```

A100 validation used:

- isolated sampler-buffer microbench over the production materialized sampler core and rowchunk helper;
- rowchunk K-sweep at fixed `R=4096`;
- correctness checks for helper extraction, greedy parity, chunk-size invariance, and non-greedy distribution sanity against a same-sampler noise floor;
- same-pod A100 serving A/B with ShareGPT traffic, using a minimal overlay identical for both backends.

Serving command shape. Row chunk sizing is exposed as an env var in this draft and can move to auto-sizing if maintainers prefer that shape:

```bash
export VLLM_DIFFUSION_GEMMA_SAMPLER_BACKEND=row_chunked
export VLLM_DIFFUSION_GEMMA_ROW_CHUNK=64
export VLLM_DIFFUSION_GEMMA_ROW_CHUNK_SCRATCH_MIB=1024

.venv/bin/python -m vllm.entrypoints.cli.main serve \
  google/diffusiongemma-26B-A4B-it \
  --host 0.0.0.0 \
  --port 8000 \
  --dtype bfloat16 \
  --gpu-memory-utilization 0.90 \
  --max-model-len 8192 \
  --max-num-seqs 16 \
  --max-num-batched-tokens 4096 \
  --trust-remote-code
```

Benchmark shape:

```bash
.venv/bin/python -m vllm.entrypoints.cli.main bench serve \
  --backend openai-chat \
  --endpoint /v1/chat/completions \
  --base-url http://127.0.0.1:8000 \
  --model google/diffusiongemma-26B-A4B-it \
  --dataset-name sharegpt \
  --dataset-path /path/to/sharegpt_first1000.json \
  --disable-shuffle \
  --num-prompts 500 \
  --sharegpt-output-len 128 \
  --ignore-eos \
  --request-rate inf \
  --max-concurrency 16 \
  --seed 123
```

## Test results

Latest tested branch head:

```text
bb80632dc478752bd7aba58cc78e76ab1f766558
```

Targeted unit/CUDA checks for the three pytest files listed above: `55 passed, 21 warnings in 47.74s`.

### Isolated sampler-buffer microbench

This microbench measures isolated sampler-buffer peak for the production materialized sampler core after helper extraction and for the rowchunk helper; it is separate from end-to-end serving VRAM.

At `R=4096`, `K=64`, current materialized peaked at `32,768 MiB`; `row_chunked` peaked at `248 MiB`, about `132x` lower for this isolated helper shape.

| sampler rows R | current materialized peak | row_chunked peak, K=64 | ratio |
|---:|---:|---:|---:|
| 256 | 2,048 MiB | 233 MiB | 8.8x |
| 1,024 | 8,192 MiB | 236 MiB | 34.7x |
| 3,072 | 24,576 MiB | 244 MiB | 100.6x |
| 4,096 | 32,768 MiB | 248 MiB | 131.9x |

K-sweep at fixed `R=4096`:

| row chunk K | row_chunked peak |
|---:|---:|
| 16 | 74 MiB |
| 32 | 132 MiB |
| 64 | 248 MiB |
| 128 | 481 MiB |
| 256 | 945 MiB |

The K-sweep is the direct shape proof for `O(K * V)` live full-vocab memory in the rowchunk path.

### Correctness checks

- Extracted materialized helper matched the original inline production-core block exactly in the small equivalence check: sampled tokens, greedy tokens, scaled logits, entropy, and soft embeddings matched.
- `row_chunked` sample and greedy token IDs were exact across tested chunk sizes: `K=16`, `32`, `64`, `128`, and `257` (`K=257` is the full-batch control for that input).
- Temperature-zero sample and greedy IDs matched between materialized and rowchunk paths with 0 mismatches.
- Non-greedy distribution sanity checks routed through `diffusion_gemma_softcap_row_chunked_sample_soft_embeds` and landed within the materialized-vs-materialized finite-sample noise-floor gate at temperatures `1.0` and `0.7`.
- Continuous side outputs can differ slightly across chunk sizes due to reduction order. The exact invariance claim is for sampled and greedy token IDs.

Distribution sanity summary, top 128 probability tokens plus one tail bin over 200,000 samples. The gate is set at `2x` the materialized-vs-materialized same-sampler control TV, so these are finite-sample noise-floor comparisons rather than absolute quality scores:

| temperature | materialized control TV | rowchunk entrypoint TV | effective TV limit | result |
|---:|---:|---:|---:|---|
| 1.0 | 0.007840 | 0.008440 | 0.015680 | pass |
| 0.7 | 0.010345 | 0.010060 | 0.020690 | pass |

### Serving pressure validation

This was a same-pod A100 validation using a minimal overlay identical for both backends, not a packaged wheel.

Dataset: ShareGPT. Output length: 128. Request rate: infinite. Shuffle disabled.

| case | backend | serving limits | traffic | readiness | completed | errors | output tok/s | total tok/s | observed max VRAM | OOM markers |
|---|---|---|---|---|---:|---:|---:|---:|---:|---:|
| lower pressure | current sampler | seqs=8, tokens=2048 | c4/o128/n200 | ready | 200/200 | 0 | 231.12 | 671.93 | 78,269 MiB | 0 |
| lower pressure | row_chunked | seqs=8, tokens=2048 | c4/o128/n200 | ready | 200/200 | 0 | 153.20 | 445.39 | 73,853 MiB | 0 |
| crossover | current sampler | seqs=8, tokens=2048 | c8/o128/n200 | ready | 8/200 | 192 | n/a | n/a | 80,745 MiB | 2 |
| crossover | row_chunked | seqs=8, tokens=2048 | c8/o128/n200 | ready | 200/200 | 0 | 176.70 | 513.71 | 73,903 MiB | 0 |
| high pressure | current sampler | seqs=16, tokens=4096 | readiness only | server died | n/a | n/a | n/a | n/a | 79,683 MiB | 4 |
| high pressure | row_chunked | seqs=16, tokens=4096 | c16/o128/n500 | ready | 500/500 | 0 | 186.87 | 559.86 | 73,947 MiB | 0 |

At lower pressure, both backends completed and the current sampler was faster. The goal of this branch is memory capacity under pressure. In the crossover case, the current sampler hit a sampler-path OOM that killed the engine; the remaining requests then failed against the downed server. At the high-pressure setting, the current sampler OOMed before reaching a usable ready state, while `row_chunked` completed. Throughput is listed as `n/a` for failed current-sampler rows because those benchmark runs did not complete.

## Known limitations

- TP/PP support is outside this PR. The rowchunk path here is TP=1 only.
- Structured output serving is not validated here.
- Quantized backends are not validated here.
- Logprobs requests fall back to the existing materialized sampler path.

## Checklist

- [x] Purpose and related work described.
- [x] Test commands included.
- [x] Test results included with separate isolated sampler-buffer and serving VRAM measurements.
- [x] AI assistance was used; the submitter reviewed the final diff and validation results.
- [ ] Documentation updates, if maintainers want an operator-facing note for the opt-in backend.

### CODE REVIEW

**Comment by masterFoad** [comment]
@mgoin @Isotr0py could you review this Draft PR when you have bandwidth?

The main review questions are:

1. Should this be kept separate from #45672, folded into #45672, or reworked as a follow-up?
2. Should the generic non-KV reserve hook be split into a smaller PR?
3. Should the draft rowchunk controls be replaced with auto-sizing, given the review feedback on #45672?

**Comment by masterFoad** [comment]
EDIT: Updated PR body, with new optimization results.

**Comment by masterFoad** [comment]
Thanks... I see #45672 landed and now bounds the DiffusionGemma sampler transient via request tiling. That covers the main OOM case this PR was originally targeting.

Before pushing a rebased update here, I wanted to ask for maintainer preference. I adapted the row-chunked path locally so it composes with the #45672 request-tiled structure: request tiles remain the outer loop, and the opt-in `row_chunked` backend bounds projection/logits/noise inside each tile while avoiding full `[rows, vocab]` logits materialization in the non-logprobs path.

Fresh local validation on A100:
- ruff: pass
- targeted pytest: `55 passed`
- helper bench, rows=256, hidden=512, vocab=262144:
  - materialized: `4.612 ms`, peak `1152 MiB`
  - row_chunked default after retune (`row_chunk=256`): `4.648 ms`, peak `771 MiB`
  - about `1.01x` latency while saving ~`381 MiB` isolated sampler peak

Smaller chunks still trade latency for more headroom, but given #45672 now handles the common case, I’m not sure the remaining opt-in backend is worth the added code surface.

Would you prefer that I:
1. push the rebased opt-in rowchunk update on top of #45672,
2. split out only any generally useful pieces,
3. split the non-KV memory-reserve hook separately, or
4. close this in favor of #45672?

Either way, I’ll update the PR body before pushing anything so it no longer overstates the pre-#45672 / `K=64` results.
